src/EnergyPlus/Cache.hh

```diff
@@ -0,0 +1,154 @@
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef ENERGYPLUS_CACHE_HH
+#define ENERGYPLUS_CACHE_HH
+
+// EnergyPlus Headers
+#include <EnergyPlus/Data/BaseData.hh>
+#include <EnergyPlus/FileSystem.hh>
+#include <EnergyPlus/IOFiles.hh>
+#include <EnergyPlus/UtilityRoutines.hh>
+
+// JSON Header
+#include <nlohmann/json.hpp>
+
+namespace EnergyPlus {
+
+// Forward declarations
+struct EnergyPlusData;
+
+namespace Cache {
+
+    std::string const cacheName = "eplusout.cache";
+    std::string const CTFKey = "CTFs";
+
+    void readJSONfile(EnergyPlusData &state, std::string &filePath, nlohmann::json &j);
+
+    void writeJSONfile(nlohmann::json &j, std::string &fPath);
+
+    template <typename T> void jsonToArray(EnergyPlusData &state, Array1D<T> &arr, nlohmann::json &j, std::string const &key)
+    {
+        // 0-based array to JSON list
+
+        try {
+            int size = static_cast<int>(j.at(key).size());
+            arr.dimension({0, size - 1});
+            int idx = 0;
+            for (auto &v : j.at(key)) {
+                arr(idx) = v;
+                ++idx;
+            }
+        } catch (const nlohmann::json::out_of_range &e) {
+            ShowFatalError(state, format("From eplusout.cache, key: \"{}\" not found", key));
+        }
+    }
+
+    template <typename T> void jsonToArray1(EnergyPlusData &state, Array1D<T> &arr, nlohmann::json &j, std::string const &key)
+    {
+        // 1-based array to JSON list
+
+        try {
+            int size = static_cast<int>(j.at(key).size());
+            arr.dimension(size);
+            int idx = 1;
+            for (auto &v : j.at(key)) {
+                arr(idx) = v;
+                ++idx;
+            }
+        } catch (const nlohmann::json::out_of_range &e) {
+            ShowFatalError(state, format("From eplusout.cache, key: \"{}\" not found", key));
+        }
+    }
+
+    template <typename T> void jsonToData(EnergyPlusData &state, T &data, nlohmann::json &j, std::string const &key)
+    {
+        try {
+            data = j.at(key);
+        } catch (const nlohmann::json::out_of_range &e) {
+            ShowFatalError(state, format("From eplusout.cache, key: \"{}\" not found", key));
+        }
+    }
+
+    template <typename T> void arrayToJSON(Array1D<T> const &arr, nlohmann::json &j, std::string const &key)
+    {
+        std::vector<T> vect;
+        for (auto v : arr)
+            vect.push_back(v);
+        j[key] = vect;
+    }
+
+    void loadCache(EnergyPlusData &state);
+
+    void writeCache(EnergyPlusData &state);
+
+    inline unsigned long long prepFloatForCacheKey(double const x, unsigned int const precision_bits)
```

Comment:
A new method for taking a floating point number to an unsigned long long with a given number of precision bits.